File: react_native_app/src/features/productos/domain/usecases/searchProductos.ts
import { UseCase } from '../../../../core/usecase/usecase';
import { Producto } from '../entities/Producto';
import { ProductoRepository } from '../repositories/ProductoRepository';

/** Busca/filtra productos por texto. */
export class SearchProductos implements UseCase<Producto[], string> {
  constructor(private readonly repository: ProductoRepository) {}

  call(query: string): Promise<Producto[]> {
    return this.repository.searchProductos(query);
  }
}

File: flutter_app/lib/features/productos/domain/usecases/search_productos.dart
import '../../../../core/usecase/usecase.dart';
import '../entities/producto.dart';
import '../repositories/producto_repository.dart';

/// Busca/filtra productos por texto.
class SearchProductos implements UseCase<List<Producto>, String> {
  final ProductoRepository repository;
  SearchProductos(this.repository);

  @override
  Future<List<Producto>> call(String query) => repository.searchProductos(query);
}
